An example mypy configuration

kaapstorm opened the pull request:
## Technical Summary

An example mypy configuration.

Context: [[CEP] The use of type hints](https://github.com/dimagi/commcare-hq/issues/31160)


## Safety Assurance

### Safety story

### Automated test coverage

This example would become part of automated checks.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

kaapstorm committed:
Add starter config for mypy

kaapstorm committed:
Add mypy to test-requirements

kaapstorm committed:
Make static type checking pass

```
$ mypy @mypy_typed_modules.txt
Success: no issues found in 1 source file
```

snopoke reviewed (commented):
I'd be interested to see a github actions job that validates this

kaapstorm committed:
Add GitHub Action for type checking

snopoke commented on .github/workflows/type_checking.yml:
```diff
@@ -0,0 +1,26 @@
+name: Type checking
+
+on:
+  push:
+    paths:
+      - '*.py'
```
won't this only trigger when the code is pushed to master?

kaapstorm replied:
It should trigger when you push to a PR. (I think `paths` should actually be `'**.py'`.)

Fwiw, you can specify to only trigger on pushes to master with
```yaml
  push:
    branches: [master]
```

kaapstorm committed:
Fix GitHub Action?

kaapstorm committed:
Merge branch 'master' into nh/mypy

kaapstorm committed:
make requirements

kaapstorm committed:
Use push event

kaapstorm committed:
Move type-checking.yml to tests.yml

kaapstorm committed:
Set config file

kaapstorm committed:
install types

kaapstorm committed:
Specify modules in module notation

kaapstorm committed:
Merge branch 'master' into nh/mypy

kaapstorm commented on corehq/motech/serializers.py:
```diff
@@ -101,7 +101,7 @@ def to_datetime_str(value):
 
 
-serializers = {
+serializers: dict[tuple[Optional[str], Optional[str]], Callable] = {
```
https://github.com/dimagi/commcare-hq/pull/31475/commits/808805d3c062da99a804c261c4a263c2d2b0a1b4

millerdev replied:
That's more verbosity and no more useful IMO.

kaapstorm replied:
If you want to use variables as parameters or default values, and you want a type checker to verify their type, then the type checker needs to know what type they are. You can choose to use this format, or the `# type: ...` comment format.

Alternatively, if you don't want the type checker to check its type, you can explicitly tell the type checker to ignore it using `# type: ignore` or (preferred) `# type: ignore[assignment]`.

kaapstorm commented on corehq/motech/value_source.py:
```diff
@@ -27,32 +30,34 @@
+    created: Optional[bool] = None
+    closed: Optional[bool] = None
+    extra_fields: dict[str, Any] = field(default_factory=dict)
+    form_question_values: dict[str, Any] = field(default_factory=dict)
```
No, the switch wasn't required, but it helps mypy to identify the types of fields. In the typing-extensions library and in Python 3.11, [PEP 681](https://peps.python.org/pep-0681/) is adding more support for authors of libraries like attrs, pydantic, SQLAlchemy and Django ORMs, etc. so that type checkers can check field types.

kaapstorm commented:
I fixed some problems with the config file, and cleaned up the commit history here: https://github.com/dimagi/commcare-hq/pull/31475

Closing this PR.
